clippy_lints/src/field_reassign_with_default.rs

```diff
@@ -0,0 +1,208 @@
+                        format!("{} {{ {}, ..Default::default() }}", binding_type, field_list)
+                    } else {
+                        format!("{} {{ {} }}", binding_type, field_list)
+                    };
+
+                    // span lint once per statement that binds default
+                    span_lint_and_note(
+                        cx,
+                        FIELD_REASSIGN_WITH_DEFAULT,
+                        first_assign.unwrap_or_else(|| unreachable!()).span,
+                        "field assignment outside of initializer for an instance created with Default::default()",
+                        Some(preceding_local.span),
+                        &format!("consider initializing the variable with `{}`", sugg),
+                    );
+                }
+            }
+        }
+    }
+}
+
+/// Returns the block indices, identifiers and types of bindings set as `Default::default()`, except
+/// for when the pattern type is a tuple.
+fn enumerate_bindings_using_default<'cx, 'hir>(
+    cx: &LateContext<'cx>,
+    block: &Block<'hir>,
+) -> Vec<(usize, Symbol, &'cx TyS<'cx>)> {
+    block
+        .stmts
+        .iter()
+        .enumerate()
+        .filter_map(|(idx, stmt)| {
+            if_chain! {
+                // only take `let ...` statements
+                if let StmtKind::Local(ref local) = stmt.kind;
+                // only take bindings to identifiers
+                if let PatKind::Binding(_, _, ident, _) = local.pat.kind;
+                // only when assigning `... = Default::default()`
+                if let Some(ref expr) = local.init;
+                if let ExprKind::Call(ref fn_expr, _) = &expr.kind;
+                if let ExprKind::Path(qpath) = &fn_expr.kind;
+                if let Res::Def(_, def_id) = qpath_res(cx, qpath, fn_expr.hir_id);
+                // right hand side of assignment is `Default::default`
+                if match_def_path(cx, def_id, &paths::DEFAULT_TRAIT_METHOD);
+                then {
+                    // Get the type of the pattern
+                    let ty = cx.typeck_results().pat_ty(local.pat);
+                    // Ignore tuples
+                    if let ty::Tuple(_) = ty.kind() {
+                        return None;
+                    }
+                    Some((idx, ident.name, ty))
+                }
+                else {
+                    None
+                }
+            }
+        })
+        .collect()
+}
+
+fn stmt_shadows_binding(this: &Stmt<'_>, shadowed: Symbol) -> bool {
+    if let StmtKind::Local(local) = &this.kind {
+        if let PatKind::Binding(_, _, ident, _) = local.pat.kind {
+            return ident.name == shadowed;
+        }
+    }
+    false
+}
+
+/// Returns the reassigned field and the assigning expression (right-hand side of assign).
+fn field_reassigned_by_stmt<'hir>(this: &Stmt<'hir>, binding_name: Symbol) -> Option<(Ident, &'hir Expr<'hir>)> {
```

Comment:
```suggestion
fn field_reassigned_by_stmt<'tcx>(this: &Stmt<'tcx>, binding_name: Symbol) -> Option<(Ident, &'tcx Expr<'tcx>)> {
```
For consistency with most of the codebase